### PR TITLE
Fix wso2/product-ei#1367

### PR DESCRIPTION
**Pull request by Manuri**
## Purpose
Fix wso2/product-ei#1367. This affects only nhttp transport.

## Goals
Fix possible thread hanging at [1]

[1] https://github.com/wso2/wso2-httpcore-nio/blob/master/modules/httpcore-nio/src/main/java/org/apache/http/nio/util/SharedOutputBuffer.java#L273

## Approach
Here SharedOutputBuffer class in httpcore-nio component has been copied inside synapse as NhttpSharedOutputBuffer and a possible thread hanging issue in SharedOutputBuffer has been fixed in NhttpSharedOutputBuffer. Class was copied instead of doing the fix in httpcore-nio component itself, was done to fix the issue without having to do an API change in httpcore-nio component

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/ClientHandler.java [hunk 1]
```diff
@@ -1395,6 +1400,7 @@ public void stop() {
     // ----------- utility methods -----------
 
     private void handleException(String msg, Exception e, NHttpClientConnection conn) {
+        log.info("Inside handleException");
```
Comment (isudana) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/ClientHandler.java]: please remove this log statement.

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+ * The I/O dispatch thread is expected to transfer data from the buffer to
+ *   {@link ContentEncoder} by calling {@link #produceContent(ContentEncoder)}.
+ * <p>
+ * The worker thread is expected to write data to the buffer by calling
+ * {@link #write(int)}, {@link #write(byte[], int, int)} or {@link #writeCompleted()}
+ * <p>
+ * In case of an abnormal situation or when no longer needed the buffer must be
+ * shut down using {@link #shutdown()} method.
+ *
+ * Please note that this class was copied from httpcore-nio in order to fix
```
Comment (isudana) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java]: Please specify which class is copied

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/ClientHandler.java (hunk 1 above)
Comment (VijithaEkanayake) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/ClientHandler.java]: Shall we remove this log?

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java (hunk 2 above)
Comment (Manuri) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java]: Will do

File: modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+                        throw new InterruptedIOException("Output operation aborted");
+                    }
+                    // Request event notifications to be triggered when the underlying
+                    // channel is ready for output operations.
+                    if (this.ioctrl != null) {
+                        this.ioctrl.requestOutput();
+                    }
+                    // Ask the thread which is reading this buffer to wait for a notification to
+                    // write data out
+                    awaitInterrupted = this.condition.await(timeout, TimeUnit.MILLISECONDS);
```
Comment (abeykoon) [inline modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java]: what happens to reading threads if awaitInterrupted=true? Do they get notified?

Comment (Manuri) [reply modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/nhttp/NhttpSharedOutputBuffer.java]: awaitInerrupted becomes true when the thread is notified.

**Pull request merged**